tests/tests/class-plugin-functions.php

```diff
@@ -0,0 +1,93 @@
+<?php
+/**
+ * Test plugin's common functions
+ *
+ * @package WP_CLI_Cron_Control_Offload
+ */
+
+namespace Automattic\WP\WP_CLI_Cron_Control_Offload\Tests;
+use Automattic\WP\WP_CLI_Cron_Control_Offload;
+use WP_UnitTestCase;
+
+/**
+ * Core function tests
+ */
+class Plugin_Functions extends WP_UnitTestCase {
+	/**
+	 * Prepare test environment
+	 */
+	function setUp() {
+		parent::setUp();
+
+		// make sure the schedule is clear.
+		_set_cron_array( array() );
+	}
+
+	/**
+	 * Clean up after our tests
+	 */
+	function tearDown() {
+		// make sure the schedule is clear.
+		_set_cron_array( array() );
+
+		parent::tearDown();
+	}
+
+	/**
+	 * Test whitelisted commands
+	 */
+	function test_whitelist_using_is_command_allowed() {
+		$this->assertTrue( WP_CLI_Cron_Control_Offload\is_command_allowed( 'post' ) );
+	}
+
+	/**
+	 * Test blacklisted commands
+	 */
+	function test_blacklist_using_is_command_allowed() {
+		$this->assertFalse( WP_CLI_Cron_Control_Offload\is_command_allowed( 'cli' ) );
+	}
+
+	/**
+	 * Test whitelisted command validation
+	 */
+	function test_whitelist_using_validate_command() {
+		$this->assertTrue( is_string( WP_CLI_Cron_Control_Offload\validate_command( 'wp post list' ) ) );
+		$this->assertTrue( is_string( WP_CLI_Cron_Control_Offload\validate_command( 'post list' ) ) );
+	}
+
+	/**
+	 * Test blacklisted command validation
+	 */
+	function test_blacklist_using_validate_command() {
+		$this->assertTrue( is_wp_error( WP_CLI_Cron_Control_Offload\validate_command( 'wp cli info' ) ) );
+		$this->assertTrue( is_wp_error( WP_CLI_Cron_Control_Offload\validate_command( 'cli info' ) ) );
+	}
+
```

Comment:
Can we add some tests for improperly constructed commands? I'm thinking things like:

* Missing `wp` (`post list`),
* Trying to sneak additional commands on (`wp post list && cat /etc/passwd`) (even though this might not actually do anything)
* Poorly constructed arguments, like `wp do-thing --foo=bar--fubar=yes`
* Backtick operators
* Trying to override `--url`, perhaps with a different site that's not part of this network
* Anything else that isn't a standard `wp command subcommand [positional_args] [--assoc_args]`

I'd like to be as defensive and paranoid as possible here.

Reply:
Good call. Pipes and ampersands should also be dropped.

The leading `wp` is optional, mostly for convenience. `WP_CLI::runcommand()` doesn't want it, so the `parse_command()` method strips it. This has the added benefit of normalizing the command to ensure cron's duplicate scheduling kicks in when appropriate.

There's a limit to how much parsing we can do without recreating large portions of WP-CLI (or loading it, in the case of custom commands), but we clearly have room for improvement. That we trust the initial users for this (requires a Go sandbox) helps somewhat.

Reply:
Checks added in 94d630e, tests in 966d0f3.

Reply:
Much beyond this is difficult, as I haven't found a way yet to invoke WP-CLI's command validation outside of where WP-CLI uses it.